### PR TITLE
Fix functional tests generated by build-init to be rerunnable 

### PR DESCRIPTION
**Pull request by tresat**
Ensure that when gradle plugin projects are setup via the build-init plugin, the generated functional tests are rerunnable using "gradle check --rerun-tasks".

Previously, the way build scripts were applied in the Spock and JUnit templates would result in duplicate plugin application blocks which would fail to run.

Fixes https://github.com/gradle/gradle/issues/18206.

### Context
This will reduce a potential point of confusion for new plugin authors.

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [ ] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

**Comment by tresat** [comment]
@bot-gradle test this

**Review by big-guy — CHANGES_REQUESTED** [review]
The tests are OK here, but we should also update PluginFunctionalTest.kt.template to follow the same kind of pattern.

The reason that the Kotlin test was working, but the Groovy/Java ones were not is the Kotlin test overwrites the file every time vs appending to it.

File: subprojects/build-init/src/main/resources/org/gradle/buildinit/tasks/templates/plugin/groovy/spock/PluginFunctionalTest.groovy.template [hunk 1]
```diff
@@ -13,12 +13,20 @@ class ${className.javaIdentifier} extends Specification {
         given:
         def projectDir = new File("build/functionalTest")
         projectDir.mkdirs()
-        new File(projectDir, "settings.gradle") << ""
-        new File(projectDir, "build.gradle") << """
+
+        def settingsFile = new File(projectDir, "settings.gradle")
```
[the remark this answers is not in the document]
Comment (tresat) [reply subprojects/build-init/src/main/resources/org/gradle/buildinit/tasks/templates/plugin/groovy/spock/PluginFunctionalTest.groovy.template:17]: Definitely an improved test setup, I was just looking to make a quick fix.

Looks like we're using JUnit/Spock/kotlintest for each flavor of functional test template, but I'll use the appropriate temp directory process for each.

**Comment by tresat** [comment]
@bot-gradle test QF

File: subprojects/build-init/src/main/resources/org/gradle/buildinit/tasks/templates/plugin/groovy/spock/PluginFunctionalTest.groovy.template [hunk 2]
```diff
@@ -3,22 +3,32 @@
  */
 ${packageDecl.statement}
 import spock.lang.Specification
+import spock.lang.TempDir
```
Comment (big-guy) [inline subprojects/build-init/src/main/resources/org/gradle/buildinit/tasks/templates/plugin/groovy/spock/PluginFunctionalTest.groovy.template:6]: Ah, good spot on this.

File: subprojects/build-init/src/main/resources/org/gradle/buildinit/tasks/templates/plugin/groovy/spock/PluginFunctionalTest.groovy.template [hunk 3]
```diff
@@ -3,22 +3,32 @@
...
-        def projectDir = new File("build/functionalTest")
-        projectDir.mkdirs()
-        new File(projectDir, "settings.gradle") << ""
-        new File(projectDir, "build.gradle") << """
-            plugins {
-                id('${pluginId.value}')
-            }
-        """
+        getSettingsFile() << ""
+        getBuildFile() << """
```
Comment (big-guy) [inline subprojects/build-init/src/main/resources/org/gradle/buildinit/tasks/templates/plugin/groovy/spock/PluginFunctionalTest.groovy.template:27]: Just so this looks a little Groovier:
```suggestion
        settingsFile << ""
        buildFile << """
```

File: subprojects/build-init/src/main/resources/org/gradle/buildinit/tasks/templates/plugin/kotlin/kotlintest/PluginFunctionalTest.kt.template [hunk 4]
```diff
@@ -11,16 +12,19 @@ import kotlin.test.assertTrue
  * A simple functional test for the '${pluginId.value}' plugin.
  */
 class ${className.javaIdentifier} {
+    private val projectDir = Files.createTempDirectory("functionalTest").toFile()
+
+    private fun getBuildFile() = projectDir.resolve("build.gradle")
+    private fun getSettingsFile() = projectDir.resolve("settings.gradle")
+
```
Comment (big-guy) [inline subprojects/build-init/src/main/resources/org/gradle/buildinit/tasks/templates/plugin/kotlin/kotlintest/PluginFunctionalTest.kt.template:19]: Ah, I was wrong that we were using JUnit Platform for Kotlin too :/

Could we use JUnit4's `TemporaryFolder` here? Something like:
```suggestion
    @get:Rule val tempFolder = TemporaryFolder()

    private fun getProjectDir() = tempFolder.root
    private fun getBuildFile() = getProjectDir().resolve("build.gradle")
    private fun getSettingsFile() = getProjectDir().resolve("settings.gradle")
```

**Review by big-guy — CHANGES_REQUESTED** [review]
Very close to being done here.  Thanks!

**Comment by tresat** [comment]
@bot-gradle test QF

**Review by big-guy — APPROVED** [review]

**Comment by big-guy** [comment]
@bot-gradle test and merge

**Comment by blindpirate** [comment]
@bot-gradle test and merge

**Comment by blindpirate** [comment]
@bot-gradle test and merge

**Review by ljacomet — APPROVED** [review]

**Comment by tresat** [comment]
@bot-gradle test and merge

**Pull request merged**